In [2]:
# Import libraries.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [3]:
# read datasets
df_credits = pd.read_csv("archive/tmdb_5000_credits.csv")
df_movies = pd.read_csv("archive/tmdb_5000_movies.csv")

Load in the datasets and check for first five rows

In [4]:
df_credits.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [5]:
df_movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

### Check the general structure of both the datasets

In [6]:
# Check the shape of the first dataset
df_credits.shape

(4803, 4)

In [7]:
# Check the shape of the second dataset
df_movies.shape

(4803, 20)

**df_credits has 4803 rows and 4 columns while df_mov has 4803 rows and 20 columns. Let's explore further**

In [8]:
# Check for the duplicates in the first dataset
sum(df_credits.duplicated())

0

In [9]:
# Check for the duplicates in the second dataset
sum(df_movies.duplicated())

0

**Both the datasets do not have any duplicate rows**

checking data type of datasets

In [10]:
# Check for the datatypes of each variable in the first dataset
df_credits.dtypes

movie_id     int64
title       object
cast        object
crew        object
dtype: object

In [11]:
# Check for the datatypes of each variable in the second dataset
df_movies.dtypes

budget                    int64
genres                   object
homepage                 object
id                        int64
keywords                 object
original_language        object
original_title           object
overview                 object
popularity              float64
production_companies     object
production_countries     object
release_date             object
revenue                   int64
runtime                 float64
spoken_languages         object
status                   object
tagline                  object
title                    object
vote_average            float64
vote_count                int64
dtype: object

`release_date` variable should be of the form datetime 

In [12]:
# Check for unique values in each variable in the first dataset
df_credits.nunique()

movie_id    4803
title       4800
cast        4761
crew        4776
dtype: int64

**While there are 4803 movies in total, there are only 4800 unique titles which suggest the presence of duplicates**


In [13]:
# Check for unique values in each variable in the second dataset
df_movies.nunique()

budget                   436
genres                  1175
homepage                1691
id                      4803
keywords                4222
original_language         37
original_title          4801
overview                4800
popularity              4802
production_companies    3697
production_countries     469
release_date            3280
revenue                 3297
runtime                  156
spoken_languages         544
status                     3
tagline                 3944
title                   4800
vote_average              71
vote_count              1609
dtype: int64

1. While there are 4803 movies in total, there are only 4800 unique titles and which suggest the presence of duplicates
2. There are 4800 titles but 4801 original titles which suggests further investigation in the 2 variables
3. There are 4802 unique popularity values instead of 4803 which suggests presence of duplicates
4. Status has only 3 unique values which needs further investigation to find if we need all the observations or some of them can be removed

In [14]:
# Check for detailed info in the first dataset
df_credits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   movie_id  4803 non-null   int64 
 1   title     4803 non-null   object
 2   cast      4803 non-null   object
 3   crew      4803 non-null   object
dtypes: int64(1), object(3)
memory usage: 150.2+ KB


In [15]:
# Check for detailed info in the second dataset
df_movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4803 non-null   int64  
 1   genres                4803 non-null   object 
 2   homepage              1712 non-null   object 
 3   id                    4803 non-null   int64  
 4   keywords              4803 non-null   object 
 5   original_language     4803 non-null   object 
 6   original_title        4803 non-null   object 
 7   overview              4800 non-null   object 
 8   popularity            4803 non-null   float64
 9   production_companies  4803 non-null   object 
 10  production_countries  4803 non-null   object 
 11  release_date          4802 non-null   object 
 12  revenue               4803 non-null   int64  
 13  runtime               4801 non-null   float64
 14  spoken_languages      4803 non-null   object 
 15  status               

**Presence of null values can be seen in some variables namely - `homepage`, `overview`, `release_date`, `runtime`, and `tagline`**

In [16]:
df_movies.status.value_counts()

status
Released           4795
Rumored               5
Post Production       3
Name: count, dtype: int64

**It can be seen that there are almost all the obseravtions which belong to `Released` status. Therefore, we need to limit the dataset to this value of status only as including other status types also makes no sense**


In [17]:
df_movies.isnull().sum()

budget                     0
genres                     0
homepage                3091
id                         0
keywords                   0
original_language          0
original_title             0
overview                   3
popularity                 0
production_companies       0
production_countries       0
release_date               1
revenue                    0
runtime                    2
spoken_languages           0
status                     0
tagline                  844
title                      0
vote_average               0
vote_count                 0
dtype: int64

**There are huge number of null counts in `homepage` followed by `tagline`. These are the variables which we might not even require in our further analysis**

##### As we saw above `original_title` and `title` unique values did not match, so we need to explore them further


In [18]:
# Check the instances where original title do not match with title column
df_movies[['original_title', 'title']].query('original_title != title')

original_title  \
97                                               シン・ゴジラ   
215                        4: Rise of the Silver Surfer   
235                         Astérix aux Jeux Olympiques   
317                                               金陵十三釵   
474                                           Évolution   
...                                                 ...   
4699                                      Lumea e a mea   
4719  Une femme mariée: Suite de fragments d'un film...   
4751                           Gabriela, Cravo e Canela   
4790                                              دایره   
4792                                                キュア   

                                       title  
97                             Shin Godzilla  
215   Fantastic 4: Rise of the Silver Surfer  
235             Asterix at the Olympic Games  
317                       The Flowers of War  
474                                Evolution  
...                                      ...  
4699                       The World Is Mine  
4719                       The Married Woman  
4751                                Gabriela  
4790                              The Circle  
4792                                    Cure  

[261 rows x 2 columns]

**So, we saw that there are 261 instances where `original_title` doesn't match with `title`. So, it's possible that number of unique values differ in both the columns. Therefore, we are on the same place as before that we need to explore `title` column further which we will do in the _Further Exploration and Cleaning_ section**

### Further Exploration and Cleaning 

#### Issues to be resolved:
1. Filter df_movies to the observations with Released status in the new dataset cl_mov which is always a good decision because we might need the original dataset at any time in the analysis

2. Drop the unnecessary columns

3. Drop the null observations in any variable

4. Explore and check title variable further for duplicates in df_movies

5. change the data type of the `release_date` to datetime in df_movies

6. Try to resolve the variables that find difficult to explore

**1. Filter df_mov to the observations with _Released_ status in the new dataset cl_mov which is always a good decision because we might need the original dataset at any time in the analysis**

In [19]:
# Filter df_movies with status 'Released'
cl_mov = df_movies[df_movies['status'] == "Released"]
cl_mov.head(2)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   

                                       homepage     id  \
0                   http://www.avatarmovie.com/  19995   
1  http://disney.go.com/disneypictures/pirates/    285   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500

In [20]:
# confirm only status is "Released"
cl_mov.status.unique()

array(['Released'], dtype=object)

**2. Drop the unnecessary columns**

In [21]:
# Drop columns not required further for analysis
cl_mov.drop(['homepage', 'keywords', 'original_language', 'original_title', 'tagline', 'overview', 'spoken_languages', 'status'], axis=1, inplace = True)
cl_mov.head(2)

/tmp/ipykernel_3368/1795847102.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cl_mov.drop(['homepage', 'keywords', 'original_language', 'original_title', 'tagline', 'overview', 'spoken_languages', 'status'], axis=1, inplace = True)


budget                                             genres     id  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  19995   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...    285   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   
1  139.082615  [{"name": "Walt Disney Pictures", "id": 2}, {"...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   

   runtime                                     title  vote_average  vote_count  
0    162.0                                    Avatar           7.2       11800  
1    169.0  Pirates of the Caribbean: At World's End           6.9        4500

In [22]:
# Check the shape again of cl_mov
cl_mov.shape

(4795, 12)

In [23]:
# Check the null counts again of cl_mov
cl_mov.isnull().sum()

budget                  0
genres                  0
id                      0
popularity              0
production_companies    0
production_countries    0
release_date            1
revenue                 0
runtime                 2
title                   0
vote_average            0
vote_count              0
dtype: int64

**3. Now, cl_mov looks much better and the null counts are so less that we can directly drop the null observations in any variable**

In [24]:
cl_mov.dropna(inplace = True)

/tmp/ipykernel_3368/4272081948.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cl_mov.dropna(inplace = True)


In [25]:
# Check the shape again of cl_mov
cl_mov.shape

(4792, 12)

**4. As we saw in the previous section, the `title` may have duplicates. So, check if there are any duplicates with respect to `title` variable**

In [26]:
# Check for duplicates in the title variable in cl_mov
cl_mov[cl_mov['title'].duplicated() == True]

budget                                             genres     id  \
2877  11000000  [{"id": 27, "name": "Horror"}, {"id": 18, "nam...   1255   
3693         0                      [{"id": 18, "name": "Drama"}]  10844   
4267   1377800  [{"id": 10751, "name": "Family"}, {"id": 12, "...   2661   

      popularity                               production_companies  \
2877   27.655270  [{"name": "Cineclick Asia", "id": 685}, {"name...   
3693    0.706355                                                 []   
4267    9.815394  [{"name": "Twentieth Century Fox Film Corporat...   

                                   production_countries release_date  \
2877      [{"iso_3166_1": "KR", "name": "South Korea"}]   2006-07-27   
3693      [{"iso_3166_1": "NZ", "name": "New Zealand"}]   2006-10-12   
4267  [{"iso_3166_1": "US", "name": "United States o...   1966-07-30   

       revenue  runtime            title  vote_average  vote_count  
2877  88489643    119.0         The Host           6.7         537  
3693         0    103.0  Out of the Blue           5.9          18  
4267         0    105.0           Batman           6.1         203

**5. Check for individual duplicate titles**

In [27]:
cl_mov[cl_mov['title'].str.contains('Out of the Blue')]

budget                         genres     id  popularity  \
3647       0  [{"id": 18, "name": "Drama"}]  39269    0.679351   
3693       0  [{"id": 18, "name": "Drama"}]  10844    0.706355   

                          production_companies  \
3647  [{"name": "Robson Street", "id": 71953}]   
3693                                        []   

                               production_countries release_date  revenue  \
3647       [{"iso_3166_1": "CA", "name": "Canada"}]   1980-05-01        0   
3693  [{"iso_3166_1": "NZ", "name": "New Zealand"}]   2006-10-12        0   

      runtime            title  vote_average  vote_count  
3647     94.0  Out of the Blue           6.5          17  
3693    103.0  Out of the Blue           5.9          18

In [28]:
cl_mov[cl_mov['title'].str.contains('The Host')]

budget                                             genres     id  \
972   44000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  72710   
2877  11000000  [{"id": 27, "name": "Horror"}, {"id": 18, "nam...   1255   

      popularity                               production_companies  \
972    42.933027  [{"name": "Nick Wechsler Productions", "id": 8...   
2877   27.655270  [{"name": "Cineclick Asia", "id": 685}, {"name...   

                                   production_countries release_date  \
972   [{"iso_3166_1": "US", "name": "United States o...   2013-03-22   
2877      [{"iso_3166_1": "KR", "name": "South Korea"}]   2006-07-27   

       revenue  runtime     title  vote_average  vote_count  
972   63327201    125.0  The Host           6.0        1817  
2877  88489643    119.0  The Host           6.7         537

In [29]:
cl_mov[cl_mov['title'].str.contains('Batman')]

budget                                             genres      id  \
9     250000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...  209112   
119   150000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...     272   
210   125000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...     415   
299   100000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...     414   
428    80000000  [{"id": 28, "name": "Action"}, {"id": 14, "nam...     364   
1359   35000000  [{"id": 14, "name": "Fantasy"}, {"id": 28, "na...     268   
3854    3500000  [{"id": 28, "name": "Action"}, {"id": 16, "nam...  142061   
4267    1377800  [{"id": 10751, "name": "Family"}, {"id": 12, "...    2661   

      popularity                               production_companies  \
9     155.790452  [{"name": "DC Comics", "id": 429}, {"name": "A...   
119   115.040024  [{"name": "DC Comics", "id": 429}, {"name": "L...   
210    50.073575  [{"name": "PolyGram Filmed Entertainment", "id...   
299    48.205606  [{"name": "Warner Bros.", "id": 6194}, {"name"...   
428    59.113174  [{"name": "PolyGram Filmed Entertainment", "id...   
1359   44.104469  [{"name": "PolyGram Filmed Entertainment", "id...   
3854   25.891466  [{"name": "DC Comics", "id": 429}, {"name": "W...   
4267    9.815394  [{"name": "Twentieth Century Fox Film Corporat...   

                                   production_countries release_date  \
9     [{"iso_3166_1": "US", "name": "United States o...   2016-03-23   
119   [{"iso_3166_1": "GB", "name": "United Kingdom"...   2005-06-10   
210   [{"iso_3166_1": "GB", "name": "United Kingdom"...   1997-06-20   
299   [{"iso_3166_1": "GB", "name": "United Kingdom"...   1995-05-31   
428   [{"iso_3166_1": "GB", "name": "United Kingdom"...   1992-06-19   
1359  [{"iso_3166_1": "GB", "name": "United Kingdom"...   1989-06-23   
3854  [{"iso_3166_1": "US", "name": "United States o...   2013-01-18   
4267  [{"iso_3166_1": "US", "name": "United States o...   1966-07-30   

        revenue  runtime                                    title  \
9     873260194    151.0       Batman v Superman: Dawn of Justice   
119   374218673    140.0                            Batman Begins   
210   238207122    125.0                           Batman & Robin   
299   336529144    121.0                           Batman Forever   
428   280000000    126.0                           Batman Returns   
1359  411348924    126.0                                   Batman   
3854          0     78.0  Batman: The Dark Knight Returns, Part 2   
4267          0    105.0                                   Batman   

      vote_average  vote_count  
9              5.7        7004  
119            7.5        7359  
210            4.2        1418  
299            5.2        1498  
428            6.6        1673  
1359           7.0        2096  
3854           7.9         419  
4267           6.1         203

**So, we observed that our intuition was wrong and there can be two or more movies with the same name**

**6. Now form a copy of the df_credits as cl_cr to explore and clean it further**

In [30]:
cl_cr = df_credits

**7. Drop the unnecessary columns in cl_cr**

In [31]:
cl_cr.drop(['crew'], axis=1, inplace=True)
cl_cr.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                                cast  
0  [{"cast_id": 242, "character": "Jake Sully", "...  
1  [{"cast_id": 4, "character": "Captain Jack Spa...

**8. Change the data type of the `release_date` to datetime**

In [32]:
cl_mov['release_date'] = pd.to_datetime(cl_mov['release_date'])
cl_mov.dtypes

/tmp/ipykernel_3368/730170391.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cl_mov['release_date'] = pd.to_datetime(cl_mov['release_date'])


budget                           int64
genres                          object
id                               int64
popularity                     float64
production_companies            object
production_countries            object
release_date            datetime64[ns]
revenue                          int64
runtime                        float64
title                           object
vote_average                   float64
vote_count                       int64
dtype: object